### ISSUE TITLE
SQL results tables: apparent autoscrolling is inaccessible to screen readers

### ISSUE THREAD

**Issue by Grahamwp** [apache/superset#24688]
When I make an SQL query with Superset, the resulting table seems to have some sort of autoscrolling features that barely work with either of my Windows screen readers, JAWS (my primary screen reader) and NVDA. They say the table has 55 rows (the number of rows in the query result) but I can barely scroll past four or five rows without either being told I'm at the bottom of the table or sometimes it autoscrolls to something like the 6th or 15th row. It barely works on NVDA and almost doesn't work at all with JAWS. I'm using the Superset instance at:
https://superset.wmcloud.org/superset/sqllab/

**Comment by wsxover** [comment]
This is probably because the result table is using [React Virtualized](http://bvaughn.github.io/react-virtualized/) which only renders "visible" rows. As such only about 5 rows are going to be present in the DOM where a screen reader can get at them. In visual user agents it is apparently trapping scroll events to then add the next rows into the DOM so they can be rendered, but this seems to not be working with the navigation events the mentioned screen readers are using.

The fix would require @bvaughn to figure out the events triggered by JAWS/NVDA and use those; or for Superset to switch to something other than React Virtualized as a framework (if so I would recommend something that uses actual table markup instead of a bunch of `div` elements tweaked up to kinda sorta behave like one).

That the result table is not accessible to screen readers is a rather critical problem IMO.

**Comment by bvaughn** [comment]
> In visual user agents it is apparently trapping scroll events to then add the next rows into the DOM so they can be rendered

Libraries like react-virtualized and react-window can be configured to over-render rows so that there are some "next" and "previous" rows (not visible, but in the DOM) for screen readers.

> (if so I would recommend something that uses actual table markup instead of a bunch of div elements tweaked up to kinda sorta behave like one).
>
> That the result table is not accessible to screen readers is a rather critical problem IMO.

So far as I understand, ARIA roles should be sufficient here (but maybe I'm mistaken).

**Comment by Grahamwp** [comment]
I don't personally know much about ARIA but I'd strongly recommend using a native HTML table here, for reliability and consistency, if nothing else. I asked a friend of mine who also uses a screen reader (primarily NVDA on Windows) and knows more about this sort of thing, @codeofdusk, and he very much agreed with me.

**Comment by codeofdusk** [comment]
Also CCing @dbjorge and @madalynrose for ARIA expertise.

**Comment by Simon818** [comment]
> The fix would require @bvaughn to figure out the events triggered by JAWS/NVDA and use those...

I can't speak to JAWS but when you have the option "Automatically set system focus to focusable elements" disabled. NVDA doesn't generate events at all. it is possible for NVDA's virtual cursor to be on a section of the page that isn't visually visible. I don't know a way around this besides simply rendering all rows in the DOM. I am also a bit skeptical about fudging an entire table structure with ARIA but I have technically seen it done in a way that is readable with NVDA, so i'm not going to definitively say it's impractical. My understanding is that when ARIA roles are defined properly, the controls should "look" like their given ARIA role and the screen reader may not even be able to tell the difference between that and native controls. The main problem I can foresee with tables is the complexity of maintaining the table itself, the headers, the rows, and the cells and having the ARIA line up well enough for screen readers to read them; however, if the framework claims to take care of that we could at least test to see whether screen readers play nice with the tale as long as all rows are rendered. It seems like a reasonable initial step.

**Comment by codeofdusk** [comment]
> "Automatically set system focus to focusable elements" disabled

This is the default (nvaccess/nvda#11190).

**Comment by rusackas** [comment]
Hi all... just wondering if there are any volunteers for this one. We'd love to see more volunteers on the [Accessibility project](https://github.com/orgs/Superset-Community-Partners/projects/1).

**Comment by rusackas** [comment]
I think this issue is now out of date, since the results table is based on AG Grid, and thus the circumstances have changed. If there are accessibility issues with the upgraded component, I would advise opening a new issue, but we can also reopen this one and add fresh context.
